docs: how to provide relation data for charm tests (#408)
canonical/charmlibs: .docs/how-to/provide-data-for-charm-tests.md

(how-to-provide-data-for-charm-tests)=

> # How to provide relation test data for charms

Reviewer: The page title isn't in sync with the title shown in the TOC. I think the version shown in the TOC is clearer:

```suggestion
# How to provide relation data for charm tests
```

This guide describes how charm interface libraries should provide sample relation data for charm unit tests.
By providing test data, charm libraries prevent charms from needing to know about the underlying relation data format.

> Instead, they only need to know about the library's public API and its testing API.

Reviewer: Very minor suggestion to avoid any confusion:

```suggestion
Instead, charms only need to know about the library's public API and its testing API.
```

Test data is published via a separate testing package for each interface library.

> The test package exposes functions that return fully populated `ops.testing.Relation` objects ready for use in state-transition tests.

Reviewer: How about making this more actively targeted at library authors. Something like:

```suggestion
As the author of an interface library, you should provide a separate testing package for your library.
The testing package should expose functions that return fully populated `ops.testing.Relation` objects ready for use in state-transition tests.
```

## Create a separate testing package

> Testing code should be distributed separately from runtime library code.

Reviewer: I feel "testing code" is potentially confusing - does it refer to how to test the library itself? How about:

```suggestion
Your library's testing package should be distributed separately from runtime library code.
```

The package should be defined in a `testing` subdirectory, like this:

```
interfaces/<name>/
├── src/charmlibs/interfaces/<name>/
│   └── __init__.py
├── pyproject.toml
└── testing/
    ├── src/charmlibs/interfaces/<name>_testing/
    │   └── __init__.py
    └── pyproject.toml
```

Use these naming conventions:

- Distribution package: `charmlibs-interfaces-<hyphenated-interface-name>-testing`
- Import package: `charmlibs.interfaces.<underscored_interface_name>_testing`

## Expose a testing extra and link the package versions

> The library and its testing package should depend on exact versions of each other.

Reviewer: I'd probably start by mentioning that the versions should be exactly the same. Something like:

```suggestion
Keep the library and testing package versions identical.
Releasing either one always implies releasing the other.
Also, the library and its testing package should depend on exact versions of each other.
```

Charms should specify the version of the library that they need in their dependencies.
In their testing dependencies, they should require the testing package, but not specify any version constraints.

> This ensures that charm tests always get testing helpers that match the library implementation.

Reviewer: The part about what charms should do is between advice for library authors. I recommend moving it to the section about example usage. (See a separate comment)

```suggestion
```

Assuming an interface library is currently on version 1.2.3, the dependencies in the library's `pyproject.toml` file should look like this:

```toml
[project.optional-dependencies]
testing = ["charmlibs-interfaces-<name>-testing==1.2.3"]

[tool.uv.sources]
charmlibs-interfaces-my-interface-testing = { path = "testing", editable = true }

[dependency-groups]
unit = [
    # If the library's unit tests use the testing package:
    "charmlibs-interfaces-<name>[testing]",
]
```

The dependencies in `testing/pyproject.toml` file should then look like this:

```toml
[project]
dependencies = [
    "charmlibs-interfaces-<name>==1.2.3",
]

[tool.uv.sources]
charmlibs-interfaces-my-interface = { path = "..", editable = true }
```

```{important}
Keep the library and testing package versions identical.
Releasing either one always implies releasing the other.

> ```

Reviewer: I recommend moving this to the top of the section. (See a separate comment)

```suggestion
```

## Implement the required testing API

> Your testing package must provide these two functions:

Reviewer: ```suggestion
Your testing package must provide these functions:
```

- `relation_for_provider`
- `relation_for_requirer`

Both functions must follow these rules:

1. The first argument must be `endpoint`, the charm's endpoint name for this relation. This argument must be required, and must be able to be provided positionally.
2. Any additional arguments must be optional and keyword-only. Defaults should represent a typical valid relation, not just empty databags.
4. Return a fully formed `ops.testing.Relation` object.
3. If the interface is request-response, include a `response` argument that defaults to `True`. If called with `response=False`, only the side of the relation that writes first should be populated.

If your interface needs helper utilities, include them in the testing package and document the intended usage.

One case where helpers are useful is if your library needs remote response data to match local request data.
For example, response payloads might need to correspond to IDs, keys, or other values generated by the charm under test.
In these cases, provide a testing API that makes this easy and explicit.

## Example usage in charm tests

> Charm developers should depend on the library's `testing` extra and import from the testing package:

Reviewer: I think it would be helpful to expand on this (shifting some info from further up) and including an example of how the charm's pyproject.toml would look. Something like:

````suggestion
Charms should specify the version of the library that they need in their dependencies.
In their testing dependencies, they should require the library's `testing` extra, but not specify any version constraints:

```toml
# Show an example of this
```

This ensures that charm tests always get testing helpers that match the library implementation.

Charm tests should then import from the testing package:
````

Author: Thanks, filling in your suggestion:
````suggestion
Charms should specify the version of the library that they need in their dependencies.
In their testing dependencies, they should require the library's `testing` extra, but not specify any version constraints:

```toml
[project]
# ...
dependencies = [
    "charmlibs.interfaces.my_interface>=1.1,<2",
    # ...
]

[dependency-groups]
unit = [
    "charmlibs.interfaces.my_interface[testing]",
]
```

This ensures that charm tests always get testing helpers that match the library implementation.

Charm tests should then import from the testing package:
````
